src/vision/status_quo/barbara_battles_buffered_streams.md

```diff
@@ -0,0 +1,216 @@
+* "Nested awaits" -- where the task blocks on an inner await while there remains other futures that are still awaiting results -- are easy to do but can cause a lot of trouble.
+* Lack of scoped futures makes it hard to spawn items into separate tasks for independent processing sometimes.
+
+### **What are the sources for this story?**
+
+This is based on the bug report [Footgun with Future Unordered](https://github.com/rust-lang/futures-rs/issues/2387) but the solution that Barbara came up with is something that was relayed by [farnz](https://github.com/farnz) vision doc writing session. [farnz] mentioned at the time that this pattern was frequently used in their codebase to work around this sort of hazard.
+
+### **Why did you choose Barbara to tell this story?**
+
+To illustrate that knowing Rust -- and even having a decent handle on async Rust's basic model -- is not enough to make it clear what is going on in this particular case.
+
+### **How would this story have played out differently for the other characters?**
+
+Woe be unto them! Identifying and fixing this bug required a lot of fluency with Rust and the async model. Alan in particular was probably relying on his understanding of async-await from other languages, which works very differently. In those languages, every async function is enqueued automatically for independent execution, so hazards like this do not arise (though this comes at a performance cost).
+
+### Besides timeouts for clients, what else could go wrong?
+
+The original bug report mentioned the possibility of deadlock:
+
+> When using an async friendly semaphore (like Tokio provides), you can deadlock yourself by having the tasks that are waiting in the `FuturesUnordered` owning all the semaphores, while having an item in a `.for_each()` block after `buffer_unordered()` requiring a semaphore.
+
+### Is there any way for Barbara to both produce and process work items simultaneously?
+
+Yes, in this case, she could've. For example, she might have written
+
+```rust
+async fn do_work(database: &Database) {
+    let work = do_select(database, FIND_WORK_QUERY).await?;
+
+    stream::iter(work)
+        .map(|item| async move {
+            let work_item = do_select(database, work_from_item(item)).await;
+            process_work_item(database, work_item).await;
+        })
+        .buffered(5)
+        .for_each(|()| std::future::ready(()))
+        .await;
+}
+```
+
+This would mean however that limit the number of work items being processed to 5. This may be what she wanted, but perhaps not. It may also be that similar problems come up in scenarios where Barbara can't so easily alter the call to `do_select` to "append" a call to `process_work_item` (for example, maybe she is in some function that takes a [`FuturesUnordered`] or `Stream` as argument).
```

Comment:
Note that it's common for us to have cases where `do_select` consumes *maybe* a kilobyte of RAM peak (as it returns a handle to work on), while `process_work_item` can consume 5 to 10 GiB peak, as it grabs compressed datablobs from an S3-like storage system, decompresses them, parses them, converts to another format, then sends them over TCP to a different service.

We want `do_select` to be allowed to run ahead to hide the latency of a database fetch, but `process_work_item` can't be allowed to run too parallel because it would consume too much RAM if it did.